### PR TITLE
Bitmap Library

### LINKED ISSUE #53: Paging Improvements
- Separate bitmap library
- Freeing pages
- Stress testing
- Detecting hardware memory limit

### LINKED ISSUE #68: Bitmap Library

### PR DESCRIPTION
**Pull request by Kfeavel**
Resolves #68 
Resolves #53

### CODE REVIEW

File: kernel/arch/i386/paging_asm.s [hunk 1]
```diff
@@ -0,0 +1,14 @@
+.section .text
+.align 4
+
+.global px_invalidate_page
+px_invalidate_page:
+#if TARGET_MACHINE >= TARGET_MACHINE_I486
```
Comment (Kfeavel) [inline kernel/arch/i386/paging_asm.s:6]: What is the purpose of this section?

File: kernel/lib/bitmap.cpp [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+}
+
+size_t bitmap_find_first_bit_clear(bitmap_t *bitmap, size_t size) {
+    for (size_t i = 0; i < size; i++) {
+        if (!((bitmap[INDEX_FROM_BIT(i)] >> OFFSET_FROM_BIT(i)) & 1)) return i;
+    }
+    return SIZE_T_MAX_VALUE;
+}
+
+/*
```
Comment (Kfeavel) [inline kernel/lib/bitmap.cpp]: Do we still need this diagram / should we clean it up to be more like actual documentation?

File: kernel/main.cpp [hunk 3]
```diff
@@ -113,11 +113,34 @@ extern "C" void px_kernel_main(const multiboot_info_t* mb_struct, uint32_t mb_ma
     px_rs232_print("\n");
     px_rs232_print(model);
     px_rs232_print("\n");
-    // Now that we're done make a joyful noise
+
+    px_kprintf(DBG_WARN "Mapping in test pages\n");
```
Comment (Kfeavel) [inline kernel/main.cpp]: Do we want to keep the tests now that we know it works?

File: kernel/arch/i386/paging_asm.s (hunk 1 above)
Comment (micahswitzer) [reply kernel/arch/i386/paging_asm.s:6]: To clear the TLB

File: kernel/lib/bitmap.cpp (hunk 2 above)
Comment (micahswitzer) [reply kernel/lib/bitmap.cpp]: We should probably get rid of it. We could keep it around for documentation if we want.

File: kernel/main.cpp (hunk 3 above)
Comment (micahswitzer) [reply kernel/main.cpp]: Maybe we could move it somewhere else

File: kernel/arch/i386/paging_asm.s (hunk 1 above)
Comment (Kfeavel) [reply kernel/arch/i386/paging_asm.s:6]: As per what we discussed, the invalidate page instruction does not exist on anything lower than i486, so this is necessary. Closing.

**Pull request merged**